File: code/ch09/go/apply_discount_test.go
// §9.2.1 — the flagship apply_discount unit-test suite, Go testing variant.

package discount

import "testing"

func TestApplyDiscount(t *testing.T) {
	cases := []struct {
		name                 string
		price, percent, want float64
	}{
		{"no discount", 100.0, 0, 100.0},
		{"half off", 100.0, 50, 50.0},
		{"rounds to cents", 9.99, 10, 8.99},
		{"full discount", 40.0, 100, 0.0},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got, err := ApplyDiscount(c.price, c.percent); err != nil || got != c.want {
				t.Errorf("got %v, %v; want %v", got, err, c.want)
			}
		})
	}
}

func TestRejectsBadPercent(t *testing.T) {
	if _, err := ApplyDiscount(100.0, 150); err == nil {
		t.Error("want error for percent 150, got nil")
	}
}

func TestFreeItemAllowed(t *testing.T) { // kills the `price <= 0` mutant
	if got, err := ApplyDiscount(0.0, 50); err != nil || got != 0.0 {
		t.Errorf("got %v, %v; want 0.0", got, err)
	}
}

File: code/ch09/ruby/apply_discount_test.rb
# §9.2.1 — the flagship apply_discount unit-test suite, Minitest variant.
require "minitest/autorun"

# Reduce price by percent (0..100). Raises ArgumentError on bad input.
def apply_discount(price, percent)
  raise ArgumentError, "price must be non-negative" if price < 0
  raise ArgumentError, "percent must be in 0..100" unless (0..100).cover?(percent)
  (price * (1 - percent / 100.0)).round(2)
end

class TestApplyDiscount < Minitest::Test
  def test_no_discount     = assert_equal(100.0, apply_discount(100.0, 0))
  def test_half_off        = assert_equal(50.0, apply_discount(100.0, 50))
  def test_rounds_to_cents = assert_equal(8.99, apply_discount(9.99, 10))
  def test_full_discount   = assert_equal(0.0, apply_discount(40.0, 100))

  def test_rejects_bad_percent
    assert_raises(ArgumentError) { apply_discount(100.0, 150) }
  end

  def test_free_item_allowed
    assert_equal(0.0, apply_discount(0.0, 50))  # kills the `price <= 0` mutant
  end
end
